### PR TITLE
Lazy evaluation for column defaults

### PR DESCRIPTION
**Pull request by zachmu**

### CODE REVIEW

File: sql/columndefault.go [hunk 1]
```diff
@@ -162,3 +181,38 @@ func (e *ColumnDefaultValue) CheckType(ctx *Context) error {
...
+func (u UnresolvedColumnDefault) Type() Type {
+	panic("UnresolvedColumnDefault is a placeholder node, but Type() was called")
+}
+
+func (u UnresolvedColumnDefault) IsNullable() bool {
+	return true
+}
+
+func (u UnresolvedColumnDefault) Eval(ctx *Context, row Row) (interface{}, error) {
+	panic("UnresolvedColumnDefault is a placeholder node, but Eval() was called")
```
Comment (Hydrocharged) [inline sql/columndefault.go:206]: Is this a panic or just an error? Something has gone wrong, but is it a panic-level something going wrong?

**Review by Hydrocharged — APPROVED** [review]
Overall looks good, but look at my comment regarding `ApplyDefaults`, as it may break integrators.

File: engine.go [hunk 2]
```diff
@@ -359,41 +359,36 @@ func (e *Engine) authCheck(ctx *sql.Context, node sql.Node) error {
 // This assumes that the given row has placeholder `nil` values for the default entries, and also that each column in a table is
 // present and in the order as represented by the schema. If no columns are given, then the given row is returned. Column indices should
 // be sorted and in ascending order, however this is not enforced.
-func ApplyDefaults(ctx *sql.Context, tblSch sql.Schema, cols []int, row sql.Row) (sql.Row, error) {
-	if len(cols) == 0 {
-		return row, nil
-	}
+func ApplyDefaults(ctx *sql.Context, tblSch sql.Schema, col int, row sql.Row) (sql.Row, error) {
```
[the remark this answers is not in the document]
Comment (zachmu) [reply engine.go]: Discussed offline, reverted this

File: sql/columndefault.go (hunk 1 above)
Comment (zachmu) [reply sql/columndefault.go:206]: It's akin to the behavior for unresolved column, if we hit this it's a very serious problem that needs attention

**Review by Hydrocharged — APPROVED** [review]

**Pull request merged**